crypto/stark-verifier-ethereum/contracts/stark_verifier.sol

```diff
@@ -5,117 +5,173 @@ import './interfaces/ConstraintInterface.sol';
+    // Profiling - 436384 gas used when the proof isn't copied into memory,
+    // making the memory copy much higher than estimates
+    function verify_proof(StarkProof memory proof, ConstraintSystem constraints) public returns (bool) {
         // Initalize the coin and constraint system
         (ProofParameters memory constraint_parameters, PublicCoin.Coin memory coin) = constraints.initalize_system(
             proof.public_inputs
         );
         // Write data to the coin and read random data from it
+        // Profiling - uses around 30k gas
         (
-            bytes32[] memory constraint_coeffiencents,
-            bytes32 oods_point,
-            bytes32[] memory oods_coefficients,
-            bytes32[] memory eval_points
+            uint256[] memory constraint_coeffiencents,
+            uint256 oods_point,
+            uint256[] memory oods_coefficients,
+            uint256[] memory eval_points
         ) = write_data_and_read_random(proof, constraint_parameters, coin);
         // Preform the proof of work check
         require(check_proof_of_work(coin, proof.pow_nonce, constraint_parameters.pow_bits), 'POW Failed');
         // Read the query indices from the coin
         uint8 eval_domain_log_size = constraint_parameters.log_trace_length + constraint_parameters.log_blowup;
         uint64[] memory queries = get_queries(coin, eval_domain_log_size, constraint_parameters.number_of_queries);
+        // Get the actual polynomial points which were commited too, and the inverses of the x_points where they were evaluated
+        // Profiling - uses 266873 gas extra for this call with data as compared to without
+        (uint256[] memory fri_top_layer, uint256 constraint_evaluated_oods_point) = constraints.constraint_calculations(
+            proof,
+            constraint_parameters,
+            queries,
+            oods_point,
+            constraint_coeffiencents
+        );
+
+        uint8 log_eval_domain_size = constraint_parameters.log_trace_length + constraint_parameters.log_blowup;
+        check_commitments(proof, constraint_parameters, queries, log_eval_domain_size);
+
+        // Profiling - 1086362 gas used for small fib before this call [includes the 250k used by the callout]
+        fri_check(proof, constraint_parameters.fri_layout, eval_points, log_eval_domain_size, queries, fri_top_layer);
+
+        check_out_of_domain_sample_result(proof, oods_point, constraint_evaluated_oods_point);
     }
 
     // This function write to the channel and reads from the channel to get the randomized data
     function write_data_and_read_random(
         StarkProof memory proof,
         ProofParameters memory constraint_parameters,
         PublicCoin.Coin memory coin
-    ) internal pure returns (bytes32[] memory, bytes32, bytes32[] memory, bytes32[] memory) {
+    ) internal pure returns (uint256[] memory, uint256, uint256[] memory, uint256[] memory) {
```

Comment:
Can we name the return variables? As it is now it's not clear what this function returns.